app/models/detailed_guide.rb

```diff
@@ -151,12 +144,63 @@ def related_mainstream_content_valid?
     end
   end
 
+
   def additional_related_mainstream_content_valid?
     if additional_related_mainstream_content_url.present? && additional_related_mainstream_content_title.blank?
       errors.add(:additional_related_mainstream_content_title, "cannot be blank if an additional related URL is given")
     end
   end
 
+  def related_mainstream_found
+    return unless related_mainstream_requested?
+    fetch_related_mainstream_content_ids
+    add_errors_for_missing_related_mainstream
+    add_errors_for_missing_additional_related_mainstream
+  end
+
+  def add_errors_for_missing_related_mainstream
+    if @content_ids.count >= 1
+      errors.add(:related_mainstream_content_url, "This mainstream content could not be found") if @content_ids[0].nil?
+    end
+  end
+
+  def add_errors_for_missing_additional_related_mainstream
+    if @content_ids.count > 1
+      errors.add(:additional_related_mainstream_content_url, "This mainstream content could not be found") if @content_ids[1].nil?
+    end
+  end
+
+  def fetch_related_mainstream_content_ids
+    base_paths = [related_mainstream_base_path, additional_related_mainstream_base_path].compact
+    if need_to_fetch_related_mainstream?(base_paths)
+      @content_ids ||= []
+      lookup_content_ids(base_paths)
+    else
+      @content_ids ||= []
+    end
+  end
+
+  def need_to_fetch_related_mainstream?(base_paths)
+    base_paths.any? && @content_ids.nil?
```

Comment:
So maybe instead of the separate method you can just have:
```ruby
@content_ids ||= lookup_content_ids(base_paths)
```

You could also memoize it keyed on the base_paths:

```ruby
def related_mainstream_content_ids
  @related_mainstream_content_ids ||= {}
  base_paths = [related_mainstream_base_path, additional_related_mainstream_base_path].compact
  @related_mainstream_content_ids[base_paths] ||= lookup_content_ids(base_paths)
end
```